### ISSUE TITLE
Double Check Locking for Migrations

### ISSUE THREAD

**Issue by SJrX** [golang-migrate/migrate#468]
**Is your feature request related to a problem? Please describe.**
We would like to have go-migrate incorporated into the application on start up. We have uncovered a failure mode, we would like to improve upon. Essentially while completely backwards compatible migrations are running the use of locking means that other containers and process that use this flow cannot start, because they require getting the lock, even though there are no changes.

**Describe the solution you'd like**
We'd like to see [double checking locking](https://en.wikipedia.org/wiki/Double-checked_locking) implemented (with an optional parameter), that allows go-migrate to pass if it is up-to-date without acquiring a lock.

**Describe alternatives you've considered**
None

**Additional context**
I'm just soliciting feedback on whether such a feature would be accepted if I submitted a PR.

**Comment by dhui** [comment]
Don't all new "instances" or versions of your application need to wait for the schema changes to be applied before running?

**Comment by SJrX** [comment]
No, we are following a practice of decoupling the db version from the application version, so will largely only apply migrations that are backwards compatible. The scenario someone brought up was the following, version X is running, and then we are going to deploy version Y, which creates a new index on an existing collection, and creates a new collection. While version Y is being migrated, we need have some load and want to spin up more containers of version X, well even though everything is fine for version X to run, those containers will block waiting for the lock.

**Comment by dhui** [comment]
Using your example:
Unfortunately, the locking currently isn't granular enough to support starting new instances of version X while blocking new instances of version Y.

If you can come up with a design, implementation, and tests for this, I'm happy to review and merge. Since this behavior is a bit riskier in nature, don't make it the default and gate it with an option.
